### ISSUE TITLE
API key exposed in app head in script tag 

### ISSUE THREAD

**Issue by joe-dematteo** [nuxt-modules/algolia#167]
Is it known that the apiKey is exposed in the head tag?

If so is there a recommended way to avoid having this apiKey seen publicly?

**Comment by Baroshem** [comment]
Hey,

Good question! 

When setting required configuration for the module, you need to pass both apiKey and application Id. These two variables are then added to public runtime config because they are needed for the algoliasearch library to work correctly on both client and server side. 

Because of that, both these variables are exposed and easily accessible on the browser.

Normally I would say that these kind of variables should be hidden somewhere on the server side as env variables but when I was looking at the official Algolia docs, I realised that in almost all examples they use the algoliasearch library, it is using very similar approach.

For example, if you inspect the request that is being sent from your browser to fetch search results, it will contain headers that are populated with the variables that I mentioned above.

I also looked at the Algolia dashboard and it seems that adminApiKey should be hidden while normal apiKey can be easily used.

**Comment by joe-dematteo** [comment]
> Hey,
> 
> Good question!
> 
> When setting required configuration for the module, you need to pass both apiKey and application Id. These two variables are then added to public runtime config because they are needed for the algoliasearch library to work correctly on both client and server side.
> 
> Because of that, both these variables are exposed and easily accessible on the browser.
> 
> Normally I would say that these kind of variables should be hidden somewhere on the server side as env variables but when I was looking at the official Algolia docs, I realised that in almost all examples they use the algoliasearch library, it is using very similar approach.
> 
> For example, if you inspect the request that is being sent from your browser to fetch search results, it will contain headers that are populated with the variables that I mentioned above.
> 
> I also looked at the Algolia dashboard and it seems that adminApiKey should be hidden while normal apiKey can be easily used.

Interesting. So I'm assuming a path to take is using SSR with Vueinstantsearch as described here?
[https://algolia.nuxtjs.org/advanced/vue-instantsearch#using-vue-instantsearch-with-ssr](https://algolia.nuxtjs.org/advanced/vue-instantsearch#using-vue-instantsearch-with-ssr)

**Comment by Baroshem** [comment]
Generally yes, but recently users have issues with setting this up.

https://github.com/nuxt-modules/algolia/discussions/173

You can also use useAsyncAlgoliaSearch composable as it will request the data on SSR and return it to you as pure html.

However, the case for apiKey could be reviewed as maybe there is a way to add the keys as private runtime config.

Maybe we could create a new configuration option named for example 'hideKeys' or 'usePrivateConfig' that when enabled would disable adding keys to public runtime config so that they can only be used in useAsyncAlgoliaSearch and SSR for VueInstantsearch and they are not exposed to head?

What do you think about this idea?

**Comment by joe-dematteo** [comment]
I think that allowing the config option to keep the keys in the private runtime config would be a good option for sure.

**Comment by Baroshem** [comment]
Ok sounds good to me as well.

Would you be interested in contributing that feature? I can provide all the help needed :)

**Comment by joe-dematteo** [comment]
Yes, I can look into integrating this feature. I'll come back if I need help.

**Comment by Lehoczky** [comment]
Hey @joe-dematteo 👋

As far as I know, you can safely expose your `apiKey`. From the [Algolia support website](https://support.algolia.com/hc/en-us/articles/18966776061329-Can-the-search-API-key-be-public):

> ### Can the search API key be public?
>
> Yes. Search keys are designed to be public. If you'd like to secure your data, you should use a [secured API key](https://www.algolia.com/doc/api-reference/api-methods/generate-secured-api-key/).
>
> You can read more about hiding your API keys [here](https://support.algolia.com/hc/en-us/articles/11163493552913).
